### PR TITLE
Add inital CDI code

### PR DESCRIPTION
**Pull request by RenaudWasTaken**
This Pull Request adds the initial code for CDI, we should then be able to import cdi/pkg from other projects with the following two functions:
- HasDevice -> Check if a Device given by a user is a CDI device
- UpdateOCISpecForDevices -> Update the OCI spec for a set of CDI device

This is missing tests for UpdateOCISpecForDevices but it's a first start.
This should also probably be optimized to cache the different files but these may come as a followup PR

### CODE REVIEW

**Comment by RenaudWasTaken** [comment]
cc @zvonkok

File: pkg/devices.go [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+	spec "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	root = "/etc/cdi"
+)
+
+var (
+	defaultLogger = logrus.New()
```
Comment (mrunalp) [inline pkg/devices.go:21]: We could avoid logging if we want to use this as a library. The callers could be returned errors and decide what to do.

File: pkg/devices.go [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+	"path/filepath"
+	"strings"
+
+	cdispec "github.com/container-orchestrated-devices/container-device-interface/specs-go"
+	spec "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	root = "/etc/cdi"
```
Comment (mrunalp) [inline pkg/devices.go:17]: We can have a `New()`  function that takes this as an option while defaulting to this directory if not set.

File: pkg/devices.go [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+
+		edits[spec.Kind] = spec
+		err = cdispec.ApplyOCIEditsForDevice(ociconfig, spec, d)
+		if err != nil {
+			return err
+		}
+	}
+
+	for _, spec := range edits {
+		cdispec.ApplyOCIEdits(ociconfig, spec)
```
Comment (kad) [inline pkg/devices.go:206]: no error checking for those?

File: pkg/devices_test.go [hunk 4]
```diff
@@ -0,0 +1,204 @@
+package pkg
+
+import (
+	"github.com/stretchr/testify/require"
+	"testing"
+
+	cdispec "github.com/container-orchestrated-devices/container-device-interface/specs-go"
+	spec "github.com/opencontainers/runtime-spec/specs-go"
```
Comment (kad) [inline pkg/devices_test.go:8]: unused import.

**Review by kad — CHANGES_REQUESTED** [review]
at least error checking on `ApplyOCIEdits` needs to be fixed.

One other thing I've noticed, we somewhere in all discussions lost the ability for individual device instance to inject env variables into OCI spec. I need to recheck other places.

File: pkg/devices.go (hunk 3 above)
Comment (kad) [inline pkg/devices.go:206]:
```suggestion
        err := cdispec.ApplyOCIEdits(ociconfig, spec)
        if err != nil {
            return err
        }
```

File: pkg/devices.go [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+	vendor := make(map[string]*cdispec.Spec)
+
+	err := filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
+		if info == nil || info.IsDir() {
+			log.Infof("CDI: Skipped directory %q", path)
+			return nil
+		}
+
+		if filepath.Ext(path) != ".json" {
+			log.Infof("CDI: Skipped non json terminated file %q", path)
```
Comment (uniemimu) [inline pkg/devices.go:39]: Looks to me this is just printing, the append is still happening below

File: pkg/devices.go (hunk 3 above)
Comment (RenaudWasTaken) [reply pkg/devices.go:206]: Good catch!

File: pkg/devices_test.go (hunk 4 above)
Comment (RenaudWasTaken) [reply pkg/devices_test.go:8]: Good catch!

File: pkg/devices.go (hunk 5 above)
Comment (RenaudWasTaken) [reply pkg/devices.go:39]: Good catch!

**Comment by RenaudWasTaken** [comment]
> One other thing I've noticed, we somewhere in all discussions lost the ability for individual device instance to inject env variables into OCI spec. I need to recheck other places.


Good comment, I'll make it part of the next PR

**Comment by umohnani8** [comment]
Late to the party, but initial code LGTM :)

**Review by kad — APPROVED** [review]
LGTM.
Side note: we later probably want to switch to https://golangci-lint.run/ instead of classical golint, but that's not urgent.

**Pull request merged**